### DISCUSSION TITLE
Importing Firedrake in WSL gives petsc error

### DISCUSSION THREAD

**Discussion by Katie795** [firedrakeproject/firedrake#3702] [Firedrake support] [answered]
Hi all,

I am having issues importing Firedrake, and I was hoping someone would be able to help. I am using a WSL with Ubuntu on Windows 11.

I have no issues installing it: the pytest 

`cd $VIRTUAL_ENV/src/firedrake
pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"`

yields

`(firedrake) katie795@katiesthinkpad:~/firedrake/src/firedrake$ pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"
================================================== test session starts ==================================================
platform linux -- Python 3.10.12, pytest-8.3.1, pluggy-1.5.0
rootdir: /home/katie795/firedrake/src/firedrake
configfile: setup.cfg
plugins: xdist-3.6.1, mpi-0.1, anyio-4.4.0, nbval-0.11.0
collected 3587 items / 3560 deselected / 2 skipped / 27 selected                                                              

tests/regression/test_dg_advection.py ....                                                                              [ 14%]
tests/regression/test_poisson_strong_bcs.py .................                                                           [ 77%]
tests/regression/test_poisson_strong_bcs_nitsche.py ....                                                                [ 92%]
tests/regression/test_stokes_mini.py ..                                                                                 [100%]

====================================================== warnings summary =======================================================
../fiat/FIAT/__init__.py:5
  /home/katie795/firedrake/src/fiat/FIAT/__init__.py:5: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================ 27 passed, 2 skipped, 3560 deselected, 1 warning in 61.60s (0:01:01) =============================
WARNING! There are options you set that were not used!
WARNING! could be spelling mistake, etc!
There is one unused database option. It is:
Option left: name:-k value: poisson_strong or stokes_mini or dg_advection source: command line`

However, when I try to run a simple module such as 

`sys.path.append("/home/katie795/firedrake/src/firedrake")
print(sys.path)
from firedrake import Function`

I get the error

`/bin/python3 /home/katie795/internship/NeuralPDE-main/src/tester.py
['/home/katie795/internship/NeuralPDE-main/src', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/home/katie795/.local/lib/python3.10/site-packages', '/usr/local/lib/python3.10/dist-packages', '/usr/lib/python3/dist-packages', '/home/katie795/firedrake/src/firedrake']
Traceback (most recent call last):
  File "/home/katie795/internship/NeuralPDE-main/src/tester.py", line 6, in <module>
    from firedrake import Function
  File "/home/katie795/firedrake/src/firedrake/firedrake/__init__.py", line 1, in <module>
    import firedrake_configuration
  File "/home/katie795/firedrake/src/firedrake/firedrake_configuration/__init__.py", line 37, in <module>
    options["with_parmetis"] = "parmetis" in petsc_packages()
  File "/home/katie795/firedrake/src/firedrake/firedrake_configuration/__init__.py", line 31, in petsc_packages
    with open(os.path.join(conf["PETSC_DIR"], conf["PETSC_ARCH"], "include", "petscconf.h"), "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-build-env-gml4avsy/normal/local/lib/python3.10/dist-packages/petsc/include/petscconf.h'`

It previously had the message 'Module not found:Firedrake' but sys.path.append fixed that.. So it looks like something has gone wrong with the PETSC directory. I have tried

- Reinstalling pip
- Reinstalling petsc4py with pip and pip3
- Reinstalling Firedrake several times
- Explicitly making 'PETSC_DIR=path/to/petsc' (this made the pytest fail)
- Configuring petsc using './configure', then 'make all check'

I have also looked at other Firedrake discussions such as https://github.com/firedrakeproject/firedrake/issues/1703 and https://github.com/firedrakeproject/firedrake/discussions/3578 but they seem to have problems that are not very similar.

Here are is my log file
[firedrake-install.log](https://github.com/user-attachments/files/16338938/firedrake-install.log)

For some strange reason, although I can see 'configure.log' and 'make.log' in my file system, when I try to attach it here it throws the error 'Path does not exist'.

Many thanks in advance! Any insight would be greatly appreciated.
Katie

**Comment by colinjcotter** [comment]
just checking whether you have started the venv?

**Reply by Katie795** [reply]
Hi, yes the venv is active

**Reply by dham** [reply]
I'm afraid that the output you posted shows that you are not using the python from the venv, which almost certainly means it's not active. This part: `/bin/python3` shows that you are using the system python rather than the one in the venv. Changing `sys.path` from within the running python won't help with this.

In a fresh terminal, activate the venv and then run `which python`. You are expecting to see `/home/katie795/firedrake/bin/python`. If you see something else then the venv hasn't successfully been activated.

**Reply by Katie795** [reply]
Hi, thank you for your reply.

You're right, I think the venv turned off (?) when pressing the run button on my module in vscode.

I tried again, being more careful, and got the expected output, but the same error. Please see below

katie795@katiesthinkpad:~$ source firedrake/bin/activate
(firedrake) katie795@katiesthinkpad:~$ which python
/home/katie795/firedrake/bin/python
(firedrake) katie795@katiesthinkpad:~$ /bin/python3 /home/katie795/internship/NeuralPDE-main/src/tester.py
['/home/katie795/internship/NeuralPDE-main/src', '/usr/lib/python310.zip', '/usr/lib/python3.10', '/usr/lib/python3.10/lib-dynload', '/home/katie795/.local/lib/python3.10/site-packages', '/usr/local/lib/python3.10/dist-packages', '/usr/lib/python3/dist-packages', '/home/katie795/firedrake/src/firedrake']
Traceback (most recent call last):
  File "/home/katie795/internship/NeuralPDE-main/src/tester.py", line 6, in <module>
    from firedrake import Function
  File "/home/katie795/firedrake/src/firedrake/firedrake/__init__.py", line 1, in <module>
    import firedrake_configuration
  File "/home/katie795/firedrake/src/firedrake/firedrake_configuration/__init__.py", line 37, in <module>
    options["with_parmetis"] = "parmetis" in petsc_packages()
  File "/home/katie795/firedrake/src/firedrake/firedrake_configuration/__init__.py", line 31, in petsc_packages
    with open(os.path.join(conf["PETSC_DIR"], conf["PETSC_ARCH"], "include", "petscconf.h"), "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-build-env-gml4avsy/normal/local/lib/python3.10/dist-packages/petsc/include/petscconf.h'

Maybe it's something to do with where I installed firedrake? (I am not super familiar with this stuff) But I had read somewhere that a venv should be installed separately to the project where it is activated. 

My code structure is
~
 -firedrake
 -myproject

Many thanks again,
Katie

**Reply by dham** [reply, marked as answer]
At this line:
```
(firedrake) katie795@katiesthinkpad:~$ /bin/python3 /home/katie795/internship/NeuralPDE-main/src/tester.py
```
You explicitly ran `/bin/python3` which overrides the venv and instead runs the python in  `/bin`. You also should not be modifying `sys.path`.


Simply run:
```
python /home/katie795/internship/NeuralPDE-main/src/tester.py
```

**Reply by Katie795** [reply]
Hi,

That now works, thank you! It seems like using the 'run' option in vscode automatically reroutes the python source.

Do you think its best to run code straight from the terminal to avoid this issue? Or do you know a quick fix?

Many thanks,
Katie

**Reply by dham** [reply]
You need to set the correct Python interpreter in visual studio code. When you have a Python file open in VS Code, you should be able to see a Python version in the bottom bar on the right. If you click on this it will give you the option to set which Python interpreter to use. If you are lucky then the Firedrake venv Python will be in the drop down list and you can choose it. If not, then click on `enter interpreter path` and then `Find...`. The Python you want in your case is `/home/katie795/firedrake/bin/python`

**Reply by Katie795** [reply]
Thank you! VSCode now runs as expected :D
